File: frameworks/shared/typescript/health-check.ts
/**
 * Health check manager for VelocityBench TypeScript/Node.js frameworks.
 *
 * Provides unified health check functionality with support for:
 * - Kubernetes liveness, readiness, and startup probes
 * - Database connectivity checks (PostgreSQL via node-postgres)
 * - Memory monitoring (process.memoryUsage())
 * - Connection pool statistics
 */

import { Pool } from 'pg';
import {
  HealthStatus,
  ProbeType,
  HealthCheck,
  HealthCheckResponse,
  HealthCheckOptions,
  DatabaseCheck,
  MemoryCheck,
  computeOverallStatus,
} from './types';

/**
 * Health check manager for Node.js frameworks.
 */
export class HealthCheckManager {
  private serviceName: string;
  private version: string;
  private environment: string;
  private database?: Pool;
  private startupDurationMs: number;
  private startTime: number;
  private cache: Map<string, { result: HealthCheckResponse; timestamp: number }>;
  private cacheTtl: number = 5000; // 5 seconds

  constructor(options: HealthCheckOptions) {
    this.serviceName = options.serviceName;
    this.version = options.version;
    this.environment = options.environment || 'development';
    this.database = options.database;
    this.startupDurationMs = options.startupDurationMs || 30000; // 30 seconds default
    this.startTime = Date.now();
    this.cache = new Map();
  }

  /**
   * Execute a health check probe.
   */
  async probe(probeType: string): Promise<HealthCheckResponse> {
    const probe = probeType.toLowerCase() as ProbeType;

    // Check cache
    const cached = this.getCachedResult(probe);
    if (cached) {
      return cached;
    }

    // Execute probe
    let result: HealthCheckResponse;
    switch (probe) {
      case ProbeType.LIVENESS:
        result = await this.livenessProbe();
        break;
      case ProbeType.READINESS:
        result = await this.readinessProbe();
        break;
      case ProbeType.STARTUP:
        result = await this.startupProbe();
        break;
      default:
        throw new Error(`Invalid probe type: ${probeType}`);
    }

    // Cache result
    this.cache.set(probe, { result, timestamp: Date.now() });

    return result;
  }

  /**
   * Get cached health check result if still valid.
   */
  private getCachedResult(probeType: ProbeType): HealthCheckResponse | null {
    const cached = this.cache.get(probeType);
    if (cached && Date.now() - cached.timestamp < this.cacheTtl) {
      return cached.result;
    }
    return null;
  }

  /**
   * Liveness probe: Is the process alive?
   *
   * Checks:
   * - Process is running
   * - Event loop is responsive
   *
   * Does NOT check database or external dependencies.
   */
  private async livenessProbe(): Promise<HealthCheckResponse> {
    const checks: { [key: string]: HealthCheck } = {};

    // Memory check (lightweight, no DB required)
    checks.memory = this.checkMemory();

    const overallStatus = computeOverallStatus(checks);

    return {
      status: overallStatus,
      timestamp: new Date().toISOString(),
      uptime_ms: this.getUptimeMs(),
      version: this.version,
      service: this.serviceName,
      environment: this.environment,
      probe_type: ProbeType.LIVENESS,
      checks,
    };
  }

  /**
   * Readiness probe: Can the service handle traffic?
   *
   * Checks:
   * - Process is running
   * - Database connection is healthy
   * - Connection pool has capacity
   * - Memory usage is acceptable
   */
  private async readinessProbe(): Promise<HealthCheckResponse> {
    const checks: { [key: string]: HealthCheck } = {};

    // Database check
    if (this.database) {
      checks.database = await this.checkDatabase();
    }

    // Memory check
    checks.memory = this.checkMemory();

    const overallStatus = computeOverallStatus(checks);

    return {
      status: overallStatus,
      timestamp: new Date().toISOString(),
      uptime_ms: this.getUptimeMs(),
      version: this.version,
      service: this.serviceName,
      environment: this.environment,
      probe_type: ProbeType.READINESS,
      checks,
    };
  }

  /**
   * Startup probe: Has initialization completed?
   *
   * Checks:
   * - Process is running
   * - Database connection established
   * - Warmup period finished
   */
  private async startupProbe(): Promise<HealthCheckResponse> {
    const checks: { [key: string]: HealthCheck } = {};

    // Database check
    if (this.database) {
      checks.database = await this.checkDatabase();
    }

    // Warmup check
    checks.warmup = this.checkWarmup();

    // Memory check
    checks.memory = this.checkMemory();

    const overallStatus = computeOverallStatus(checks);

    return {
      status: overallStatus,
      timestamp: new Date().toISOString(),
      uptime_ms: this.getUptimeMs(),
      version: this.version,
      service: this.serviceName,
      environment: this.environment,
      probe_type: ProbeType.STARTUP,
      checks,
    };
  }

  /**
   * Check database connectivity and pool health.
   */
  private async checkDatabase(): Promise<DatabaseCheck> {
    if (!this.database) {
      return {
        status: HealthStatus.DOWN,
        error: 'Database pool not initialized',
      };
    }

    const startTime = Date.now();

    try {
      // Execute simple query to verify connectivity
      const result = await Promise.race([
        this.database.query('SELECT 1 as health_check'),
        new Promise((_, reject) =>
          setTimeout(() => reject(new Error('Timeout')), 3000)
        ),
      ]);

      const responseTime = Date.now() - startTime;

      // Get pool statistics
      const poolSize = this.database.totalCount;
      const poolIdle = this.database.idleCount;
      const poolWaiting = this.database.waitingCount;
      const poolMaxSize = (this.database as any).options?.max || 10;

      // Calculate utilization
      const utilization = (poolSize / poolMaxSize) * 100;

      let status = HealthStatus.UP;
      let warning: string | undefined;

      if (utilization > 95) {
        status = HealthStatus.DEGRADED;
        warning = `Connection pool nearly exhausted (${poolSize}/${poolMaxSize})`;
      } else if (utilization > 80) {
        warning = `High connection pool utilization (${utilization.toFixed(1)}%)`;
      }

      return {
        status,
        response_time_ms: Math.round(responseTime * 100) / 100,
        pool_size: poolSize,
        pool_available: poolIdle,
        pool_max_size: poolMaxSize,
        warning,
      };
    } catch (error: any) {
      if (error.message === 'Timeout') {
        return {
          status: HealthStatus.DOWN,
          error: 'Database query timeout (>3s)',
        };
      }
      return {
        status: HealthStatus.DOWN,
        error: `Database connection error: ${error.message}`,
      };
    }
  }

  /**
   * Check memory usage.
   */
  private checkMemory(): MemoryCheck {
    try {
      const memUsage = process.memoryUsage();
      const rssMb = memUsage.rss / 1024 / 1024;
      const heapUsedMb = memUsage.heapUsed / 1024 / 1024;
      const heapTotalMb = memUsage.heapTotal / 1024 / 1024;

      // Estimate total system memory (approximation)
      // In production, could use os.totalmem() if needed
      const totalMb = heapTotalMb * 4; // Rough estimate
      const utilization = (rssMb / totalMb) * 100;

      let status = HealthStatus.UP;
      let warning: string | undefined;

      if (utilization > 90) {
        status = HealthStatus.DEGRADED;
        warning = `Critical memory usage (${utilization.toFixed(1)}%)`;
      } else if (utilization > 80) {
        warning = `High memory usage (${utilization.toFixed(1)}%)`;
      }

      return {
        status,
        used_mb: Math.round(rssMb * 100) / 100,
        total_mb: Math.round(totalMb * 100) / 100,
        utilization_percent: Math.round(utilization * 100) / 100,
        warning,
      };
    } catch (error: any) {
      return {
        status: HealthStatus.DEGRADED,
        warning: `Memory check error: ${error.message}`,
      };
    }
  }

  /**
   * Check if warmup period has completed.
   */
  private checkWarmup(): HealthCheck {
    const uptimeMs = this.getUptimeMs();

    if (uptimeMs < this.startupDurationMs) {
      const progress = (uptimeMs / this.startupDurationMs) * 100;
      return {
        status: HealthStatus.IN_PROGRESS,
        info: `Warming up (${progress.toFixed(0)}% complete)`,
        progress_percent: Math.round(progress * 10) / 10,
        uptime_ms: uptimeMs,
        target_ms: this.startupDurationMs,
      };
    } else {
      return {
        status: HealthStatus.UP,
        info: 'Warmup complete',
      };
    }
  }

  /**
   * Get service uptime in milliseconds.
   */
  private getUptimeMs(): number {
    return Date.now() - this.startTime;
  }
}

File: frameworks/shared/go/health_check.go
package healthcheck

import (
	"context"
	"database/sql"
	"fmt"
	"runtime"
	"sync"
	"time"
)

// HealthCheckManager manages health checks for a VelocityBench Go framework.
type HealthCheckManager struct {
	serviceName       string
	version           string
	environment       string
	database          *sql.DB
	startupDurationMs int64
	startTime         time.Time
	cache             sync.Map
	cacheTTL          time.Duration
}

type cachedResult struct {
	result    *HealthCheckResponse
	timestamp time.Time
}

// Config holds configuration for the health check manager.
type Config struct {
	ServiceName       string
	Version           string
	Environment       string
	Database          *sql.DB
	StartupDurationMs int64 // Warmup period in milliseconds
}

// New creates a new HealthCheckManager.
func New(config Config) *HealthCheckManager {
	if config.Environment == "" {
		config.Environment = "development"
	}
	if config.StartupDurationMs == 0 {
		config.StartupDurationMs = 30000 // 30 seconds default
	}

	return &HealthCheckManager{
		serviceName:       config.ServiceName,
		version:           config.Version,
		environment:       config.Environment,
		database:          config.Database,
		startupDurationMs: config.StartupDurationMs,
		startTime:         time.Now(),
		cacheTTL:          5 * time.Second,
	}
}

// Probe executes a health check probe.
func (m *HealthCheckManager) Probe(probeType string) (*HealthCheckResponse, error) {
	var probe ProbeType
	switch probeType {
	case "liveness":
		probe = ProbeLiveness
	case "readiness":
		probe = ProbeReadiness
	case "startup":
		probe = ProbeStartup
	default:
		return nil, fmt.Errorf("invalid probe type: %s", probeType)
	}

	// Check cache
	if cached := m.getCachedResult(probe); cached != nil {
		return cached, nil
	}

	// Execute probe
	var result *HealthCheckResponse
	switch probe {
	case ProbeLiveness:
		result = m.livenessProbe()
	case ProbeReadiness:
		result = m.readinessProbe()
	case ProbeStartup:
		result = m.startupProbe()
	}

	// Cache result
	m.cache.Store(string(probe), cachedResult{
		result:    result,
		timestamp: time.Now(),
	})

	return result, nil
}

// getCachedResult retrieves a cached health check result if still valid.
func (m *HealthCheckManager) getCachedResult(probeType ProbeType) *HealthCheckResponse {
	if cached, ok := m.cache.Load(string(probeType)); ok {
		if cr, ok := cached.(cachedResult); ok {
			if time.Since(cr.timestamp) < m.cacheTTL {
				return cr.result
			}
		}
	}
	return nil
}

// livenessProbe checks if the process is alive.
//
// Checks:
// - Process is running
// - Memory is accessible
//
// Does NOT check database or external dependencies.
func (m *HealthCheckManager) livenessProbe() *HealthCheckResponse {
	checks := make(map[string]interface{})

	// Memory check (lightweight, no DB required)
	checks["memory"] = m.checkMemory()

	overallStatus := m.computeOverallStatus(checks)

	return &HealthCheckResponse{
		Status:      overallStatus,
		Timestamp:   time.Now().UTC(),
		UptimeMs:    m.getUptimeMs(),
		Version:     m.version,
		Service:     m.serviceName,
		Environment: m.environment,
		ProbeType:   ProbeLiveness,
		Checks:      checks,
	}
}

// readinessProbe checks if the service can handle traffic.
//
// Checks:
// - Process is running
// - Database connection is healthy
// - Memory usage is acceptable
func (m *HealthCheckManager) readinessProbe() *HealthCheckResponse {
	checks := make(map[string]interface{})

	// Database check
	if m.database != nil {
		checks["database"] = m.checkDatabase()
	}

	// Memory check
	checks["memory"] = m.checkMemory()

	overallStatus := m.computeOverallStatus(checks)

	return &HealthCheckResponse{
		Status:      overallStatus,
		Timestamp:   time.Now().UTC(),
		UptimeMs:    m.getUptimeMs(),
		Version:     m.version,
		Service:     m.serviceName,
		Environment: m.environment,
		ProbeType:   ProbeReadiness,
		Checks:      checks,
	}
}

// startupProbe checks if initialization is complete.
//
// Checks:
// - Process is running
// - Database connection established
// - Warmup period finished
func (m *HealthCheckManager) startupProbe() *HealthCheckResponse {
	checks := make(map[string]interface{})

	// Database check
	if m.database != nil {
		checks["database"] = m.checkDatabase()
	}

	// Warmup check
	checks["warmup"] = m.checkWarmup()

	// Memory check
	checks["memory"] = m.checkMemory()

	overallStatus := m.computeOverallStatus(checks)

	return &HealthCheckResponse{
		Status:      overallStatus,
		Timestamp:   time.Now().UTC(),
		UptimeMs:    m.getUptimeMs(),
		Version:     m.version,
		Service:     m.serviceName,
		Environment: m.environment,
		ProbeType:   ProbeStartup,
		Checks:      checks,
	}
}

// checkDatabase checks database connectivity and health.
func (m *HealthCheckManager) checkDatabase() HealthCheck {
	if m.database == nil {
		errMsg := "Database not initialized"
		return HealthCheck{
			Status: StatusDown,
			Error:  &errMsg,
		}
	}

	start := time.Now()
	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
	defer cancel()

	// Execute simple query to verify connectivity
	var result int
	err := m.database.QueryRowContext(ctx, "SELECT 1").Scan(&result)

	if err != nil {
		if ctx.Err() == context.DeadlineExceeded {
			errMsg := "Database query timeout (>3s)"
			return HealthCheck{
				Status: StatusDown,
				Error:  &errMsg,
			}
		}
		errMsg := fmt.Sprintf("Database connection error: %v", err)
		return HealthCheck{
			Status: StatusDown,
			Error:  &errMsg,
		}
	}

	responseTime := float64(time.Since(start).Milliseconds())

	// Get pool statistics (if available)
	stats := m.database.Stats()
	poolSize := stats.OpenConnections
	poolMaxSize := stats.MaxOpenConnections
	poolIdle := stats.Idle

	// Calculate utilization
	var utilization float64
	if poolMaxSize > 0 {
		utilization = (float64(poolSize) / float64(poolMaxSize)) * 100
	}

	status := StatusUp
	var warning *string

	if utilization > 95 {
		status = StatusDegraded
		warnMsg := fmt.Sprintf("Connection pool nearly exhausted (%d/%d)", poolSize, poolMaxSize)
		warning = &warnMsg
	} else if utilization > 80 {
		warnMsg := fmt.Sprintf("High connection pool utilization (%.1f%%)", utilization)
		warning = &warnMsg
	}

	return HealthCheck{
		Status:         status,
		ResponseTimeMs: &responseTime,
		Warning:        warning,
		AdditionalData: map[string]interface{}{
			"pool_size":      poolSize,
			"pool_available": poolIdle,
			"pool_max_size":  poolMaxSize,
		},
	}
}

// checkMemory checks memory usage.
func (m *HealthCheckManager) checkMemory() HealthCheck {
	var memStats runtime.MemStats
	runtime.ReadMemStats(&memStats)

	usedMB := float64(memStats.Alloc) / 1024 / 1024
	totalMB := float64(memStats.Sys) / 1024 / 1024
	utilization := (usedMB / totalMB) * 100

	status := StatusUp
	var warning *string

	if utilization > 90 {
		status = StatusDegraded
		warnMsg := fmt.Sprintf("Critical memory usage (%.1f%%)", utilization)
		warning = &warnMsg
	} else if utilization > 80 {
		warnMsg := fmt.Sprintf("High memory usage (%.1f%%)", utilization)
		warning = &warnMsg
	}

	return HealthCheck{
		Status:  status,
		Warning: warning,
		AdditionalData: map[string]interface{}{
			"used_mb":             usedMB,
			"total_mb":            totalMB,
			"utilization_percent": utilization,
		},
	}
}

// checkWarmup checks if warmup period has completed.
func (m *HealthCheckManager) checkWarmup() HealthCheck {
	uptimeMs := m.getUptimeMs()

	if uptimeMs < m.startupDurationMs {
		progress := (float64(uptimeMs) / float64(m.startupDurationMs)) * 100
		infoMsg := fmt.Sprintf("Warming up (%.0f%% complete)", progress)
		return HealthCheck{
			Status: StatusInProgress,
			Info:   &infoMsg,
			AdditionalData: map[string]interface{}{
				"progress_percent": progress,
				"uptime_ms":        uptimeMs,
				"target_ms":        m.startupDurationMs,
			},
		}
	}

	infoMsg := "Warmup complete"
	return HealthCheck{
		Status: StatusUp,
		Info:   &infoMsg,
	}
}

// computeOverallStatus computes overall health status from checks.
func (m *HealthCheckManager) computeOverallStatus(checks map[string]interface{}) HealthStatus {
	hasDown := false
	hasInProgress := false
	hasDegraded := false

	for _, check := range checks {
		if hc, ok := check.(HealthCheck); ok {
			switch hc.Status {
			case StatusDown:
				hasDown = true
			case StatusInProgress:
				hasInProgress = true
			case StatusDegraded:
				hasDegraded = true
			}
		}
	}

	if hasDown {
		return StatusDown
	} else if hasInProgress {
		return StatusInProgress
	} else if hasDegraded {
		return StatusDegraded
	}
	return StatusUp
}

// getUptimeMs returns service uptime in milliseconds.
func (m *HealthCheckManager) getUptimeMs() int64 {
	return time.Since(m.startTime).Milliseconds()
}
